3.2/feature/validation changes

zeelot opened the pull request:
- #3424 http://dev.kohanaframework.org/issues/3424
- #4062 http://dev.kohanaframework.org/issues/4062
- #4034 http://dev.kohanaframework.org/issues/4034
- #3552 http://dev.kohanaframework.org/issues/3552

As many commits as possible so we can still debate things if anyone complains, but I did confirm all the changes with @zombor before applying :)

zeelot committed:
removing Validation::$_empty_rules(refs #3552)

zeelot committed:
removing _empty_rules from validation tests (refs #3552)

zeelot committed:
adding tests for all the valid helpers that fail on empty fields (refs #3552)

zeelot committed:
fixing whitespace issues

zeelot committed:
fixing one of the new tests (refs #3552)

zeelot committed:
adjusting the valid methods to pass the new tests (refs #3552)

zeelot committed:
fixing test for exact_length rule (refs #3552)

zeelot committed:
implementing ArrayAccess instead of extending ArrayObject (refs #4034)

zeelot committed:
added Validation::data() deprecated Validation::as_array() (refs #4062)

zeelot committed:
adding :data bound value (refs #4062)

zeelot committed:
updating tests to use non-deprecated Validation::data() (refs #4062)

zeelot committed:
adding tests for Validation::data() and Validation::as_array() (refs #4062)

zeelot committed:
Allowing Valid::exact_length() to validate an array of lengths (refs #3424)

zeelot committed:
testing new Valid::exact_length() format (refs #3424)

zeelot committed:
Adding another test for Valid::exact_length()

zeelot commented:
Ah, forgot to make ArrayAccess read-only and unit test the methods... I will take care of this tonight.

cbandy commented on classes/kohana/validation.php:
```diff
@@ -63,19 +119,31 @@ class Kohana_Validation extends ArrayObject {
+	}
+
+	/**
+	 * Returns the array representation of the current object.
```
This is not clear enough for me. It should say something about the data/values.

zeelot replied:
How about: `Returns the array of data to be validated.`

cbandy replied:
I like it.

zeelot committed:
Revert "adjusting the valid methods to pass the new tests (refs #3552)"

This reverts commit 3760e1d717ec162590ec54dfa9f03fbb1d9b4090.

Conflicts:

	classes/kohana/valid.php

zeelot committed:
Revert "removing _empty_rules from validation tests (refs #3552)"

This reverts commit d0e5bcd36161d214965d4c45a569078fd6606bec.

zeelot committed:
Revert "removing Validation::$_empty_rules(refs #3552)"

This reverts commit a294de50d4f46b66eab7a96f561dd8f39f2bc9e1.

zeelot committed:
adjusting the Valid tests on empty fields (refs #3552)

zeelot committed:
Valid::email_domain() doesn't need to run checkdnsrr() on empty fields (refs #3552)

zeelot committed:
Adding tests for Validation's implementation of ArrayAccess (refs #4034)

zeelot committed:
Making the ArrayAccess methods in Validation read-only (refs #4034)

zeelot committed:
updating a docblock.

zeelot committed:
Adding some extra Valid tests

zeelot commented:
This should be everything again :)
- reverted back to `_empty_rules`
- reverted all the `Valid` methods to make more sense
- fixed an issue with `Valid::email_domain` running on empty values
- added tests for `Validation` `ArrayAccess` methods
- made `Validation` read-only (`ArrayAccess`)
- added tests for `Valid` methods to run on `NULL` and `FALSE` values
